pkg/systemd/services/services.jsx

```diff
@@ -266,51 +273,61 @@ class ServicesPageBody extends React.Component {
+                return;
+            if (!path.endsWith("service") && !path.endsWith("timer") && !path.endsWith("socket"))
                 return;
 
-            if (this.state.unit_by_path[path] &&
-                this.state.unit_by_path[path].Transient &&
-                args[1].ActiveState &&
-                ((args[1].ActiveState.v == 'failed' && this.state.unit_by_path[path].CollectMode == 'inactive-or-failed') ||
-                  args[1].ActiveState.v == 'inactive')) {
-                this.seenPaths.delete(path);
-                const copy_unit_by_path = { ...this.state.unit_by_path };
-                delete copy_unit_by_path[path];
-                this.setState({
-                    unit_by_path: copy_unit_by_path,
-                });
+            const unit_id = this.unit_by_path[path];
+            const unit = this.state.units[unit_id];
+
+            if (unit) {
+                // update existing unit state
+                const newUnit = { ...unit };
+                for (const prop of ["ActiveState", "LoadState", "Description"]) {
+                    if (args[1][prop])
+                        newUnit[prop] = args[1][prop].v;
+                }
+
+                this.setState({ units: { ...this.state.units, [unit_id]: newUnit } });
+                this.processFailedUnits();
             } else {
-                this.updateProperties(args[1], path);
+                // this happens when starting an unloaded unit
+                this.listUnits();
             }
-            this.processFailedUnits();
         });
 
-        ["JobNew", "JobRemoved"].forEach(signalName => {
-            systemd_client[this.props.owner].subscribe({ interface: SD_MANAGER, member: signalName }, (path, iface, signal, args) => {
-                const unit_id = args[2];
-                systemd_client[this.props.owner].call(SD_OBJ, SD_MANAGER, "LoadUnit", [unit_id])
-                        .then(([path]) => {
-                            if (!this.seenPaths.has(path))
-                                this.seenPaths.add(path);
+        // handle transient units
+        systemd_client[this.props.owner].subscribe({ interface: SD_MANAGER, member: "UnitRemoved" }, (_path, _iface, _signal, [id, objpath]) => {
+            // during daemon reload we get tons of these, ignore
+            if (this.reloading)
+                return;
 
-                            this.getUnitByPath(path).then(this.processFailedUnits);
-                        });
-            });
+            // be careful to not remove units from ListUnitFiles(), only loaded ones; we really only need this for transient ones
+            if (this.state.units[id]?.UnitFileState === 'transient') {
+                const units = { ...this.state.units };
+                delete units[id];
+                delete this.unit_by_path[objpath];
+                this.setState({ units });
+                this.processFailedUnits(units);
```

Comment:
The `units` argument here seems to belong into the commit for fixing failed unit processing, no?

Reply:
Yep! DJ Drum & Rebase! :headphones: :notes: